projects/sandbox/train/train/model.py

```diff
@@ -0,0 +1,209 @@
+    def forward(self, parameters, strain) -> Tensor:
+        return -self.model.log_prob(parameters, context=strain)
+
+    def training_step(self, batch, _):
+        strain, parameters = batch
+        loss = self(parameters, strain).mean()
+        self.log(
+            "train_loss",
+            loss,
+            on_step=True,
+            on_epoch=True,
+            prog_bar=True,
+            sync_dist=False,
+            logger=True,
+        )
+        return loss
+
+    def validation_step(self, batch, _):
+        strain, parameters = batch
+        loss = self(parameters, strain).mean()
+        self.log(
+            "valid_loss",
+            loss,
+            on_epoch=True,
+            prog_bar=True,
+            sync_dist=True,
+            logger=True,
+        )
+        return loss
+
+    def cast_as_bilby_result(
+        self,
+        samples: np.ndarray,
+        truth: np.ndarray,
+    ):
+        """Cast posterior samples as Bilby Result object
+        for ease of producing corner and pp plots
+
+        Args:
+            samples: posterior samples (1, num_samples, num_params)
+            truth: true values of the parameters  (1, num_params)
+            priors: dictionary of prior objects
+            label: label for the bilby result object
+
+        """
+
+        inference_params = self.trainer.datamodule.inference_params
+        injection_parameters = {
+            k: float(v) for k, v in zip(inference_params, truth)
+        }
+
+        # create dummy prior with correct attributes
+        # for making our results compatible with bilbys make_pp_plot
+        priors = {
+            param: bilby.core.prior.base.Prior(latex_label=param)
+            for param in inference_params
+        }
+        posterior = dict()
+        for idx, k in enumerate(inference_params):
+            posterior[k] = samples.T[idx].flatten()
+        posterior = pd.DataFrame(posterior)
+
+        return Result(
+            label="PEModel",
+            injection_parameters=injection_parameters,
+            posterior=posterior,
+            search_parameter_keys=inference_params,
+            priors=priors,
+        )
+
+    def on_test_epoch_start(self):
+        self.test_results = []
+        self.num_plotted = 0
+
+    def test_step(self, batch, batch_idx):
+        # whitened strain and de-scaled parameters
+        strain, parameters = batch
+        samples = self.model.sample(
+            self.hparams.num_samples_draw, context=strain
+        )
+        descaled = self.trainer.datamodule.scale(samples, reverse=True)
+        result = self.cast_as_bilby_result(
+            descaled.cpu().numpy(),
+            parameters.cpu().numpy()[0],
+        )
+        self.test_results.append(result)
+
+        if batch_idx % 100 == 0 and self.num_plotted < self.num_corner:
+            skymap_filename = self.outdir / f"{self.num_plotted}_mollview.png"
+            corner_filename = self.outdir / f"{self.num_plotted}_corner.png"
```

Comment:
Let's also dump the result file as json or pickle for the cases where we have the mollview and corner plots.